src/libraries/System.Security.Cryptography/tests/OpenSslNamedKeysTests.manual.cs

```diff
@@ -174,17 +225,202 @@ public static void Engine_UsePublicKey()
+            ecdsaPub.ImportSubjectPublicKeyInfo(ecdsaPubBytes, out int bytesRead);
+            Assert.Equal(ecdsaPubBytes.Length, bytesRead);
+
+            using ECDsa ecdsaBad = ECDsa.Create();
+            ecdsaBad.KeySize = ecdsaPri.KeySize;
+
+            // Verify can sign/verify multiple times
+            for (int i = 0; i < 10; i++)
+            {
+                data[0] = (byte)i;
+                byte[] signature = ecdsaPri.SignData(data, HashAlgorithmName.SHA256);
+                byte[] badSignature = ecdsaBad.SignData(data, HashAlgorithmName.SHA256);
+                Assert.NotEqual(data, signature);
+                Assert.NotEqual(data, badSignature);
+                Assert.NotEqual(badSignature, signature);
+                Assert.True(ecdsaPub.VerifyData(data, signature, HashAlgorithmName.SHA256));
+                Assert.False(ecdsaPub.VerifyData(data, badSignature, HashAlgorithmName.SHA256));
+                Assert.False(ecdsaBad.VerifyData(data, signature, HashAlgorithmName.SHA256));
+
+                // TPM key is intended for sign/decrypt only, we could theoretically make verify work without needing to export/import by forcing 'default' provider
+                // for this operation but it's most likely misusage on user part and tpm2 provider intentionally didn't allow it so we will follow this logic.
+                Assert.ThrowsAny<CryptographicException>(() => ecdsaPri.VerifyData(data, signature, HashAlgorithmName.SHA256));
+            }
+
+            // It's TPM so it should not be possible to export parameters
+            Assert.ThrowsAny<CryptographicException>(() => ecdsaPri.ExportParameters(includePrivateParameters: true));
+        }
+
+        [ConditionalFact(nameof(ShouldRunProviderEcDhTests))]
+        public static void Provider_TPM2ECDH()
+        {
+            using SafeEvpPKeyHandle priKeyHandle = SafeEvpPKeyHandle.OpenKeyFromProvider(Tpm2ProviderName, TpmEcDhKeyHandleUri);
+            using ECDiffieHellman alicePri = new ECDiffieHellmanOpenSsl(priKeyHandle);
+            using ECDiffieHellman alicePub = ECDiffieHellman.Create();
+            alicePub.ImportParameters(alicePri.ExportParameters(includePrivateParameters: false));
+
+            using ECDiffieHellman bobPri = ECDiffieHellman.Create();
+            bobPri.KeySize = alicePri.KeySize;
+
+            using ECDiffieHellman bobPub = ECDiffieHellman.Create();
+            bobPub.ImportParameters(bobPri.ExportParameters(includePrivateParameters: false));
+
+            byte[] sharedKeyFromAlice;
+            using (ECDiffieHellmanPublicKey bobPublic = bobPub.PublicKey)
+            {
+                sharedKeyFromAlice = alicePri.DeriveKeyMaterial(bobPublic);
+
+                Assert.NotEmpty(sharedKeyFromAlice);
+
+                byte firstByte = sharedKeyFromAlice[0];
+                bool allSame = sharedKeyFromAlice.All((x) => x == firstByte);
+                Assert.False(allSame, "all bytes of shared key are the same");
```

Comment:
Is this saying that `DeriveKeyMaterial` on a key from TPM2 is going to produce an output like `01010101010101` or `02020202020202`, just one byte repeated over and over again?

That's weird, and bad.  (Though, since I believe the DeriveKeyMaterial you're calling to be doing SHA256(z), that's really impressive if it's true)

Reply:
It's mostly sanity check against us doing literally nothing and just returning zeros or 0xff everywhere (both sides are happy because secrets match but not exactly what they wanted)

Reply:
I changed that to raw secret but leaving the all bytes same check (that's a red flag something is really wrong and we might have not done anything)